File: Flyweight/DrawingForm.cs
namespace Flyweight
{
    public partial class DrawingForm : Form
    {
        private readonly int WIDTH = 818;
        private readonly int HEIGHT = 395;
        private static readonly ShapeType[] shapes = { ShapeType.LINE, ShapeType.OVAL_FILL, ShapeType.OVAL_NOFILL };
        private static readonly Color[] colors = { Color.Red, Color.Green, Color.Orange };
        private readonly Random rand = new Random();
        private bool stopRequested = false;
        private bool isDrawing = false;
        private List<IShape> shapeList = new List<IShape>();
        public static DrawingForm? Instance;

        public DrawingForm()
        {
            InitializeComponent();
            Instance = this;
            drawTimer.Interval = 2000;
        }

        public void ClearMessage()
        {
            textBoxOutput.Clear();
        }

        public void ShowMessage(string message)
        {
            textBoxOutput.AppendText(message);
        }

        private void BtnStart_Click(object sender, EventArgs e)
        {
            if (!isDrawing)
            {
                stopRequested = false;
                shapeList.Clear();
                isDrawing = true;
                drawTimer.Start();
            }
        }

        private void BtnStop_Click(object sender, EventArgs e)
        {
            stopRequested = true;
        }

        private ShapeType GetRandomShape()
        {
            return shapes[rand.Next(0, shapes.Length)];
        }

        private Color GetRandomColor()
        {
            return colors[rand.Next(0, colors.Length)];
        }

        private void pictrBox_Paint(object sender, PaintEventArgs e)
        {
            if (stopRequested || shapeList.Count >= 20)
            {
                ClearMessage();
                isDrawing = false;
                return;
            }

            Graphics g = e.Graphics;
            foreach (IShape shape in shapeList)
            {
                shape.Draw(g);
            }
        }

        private void drawTimer_Tick(object sender, EventArgs e)
        {
            if (!stopRequested && shapeList.Count < 20)
            {
                IShape shape = ShapeFactory.GetShape(GetRandomShape(), GetRandomColor(), GetRandomX(), GetRandomY(), GetRandomWidth(), GetRandomHeight());
                shapeList.Add(shape);
                pictrBox.Invalidate();
            }
        }

        private int GetRandomX()
        {
            return rand.Next(0, WIDTH);
        }

        private int GetRandomY()
        {
            return rand.Next(0, HEIGHT);
        }

        private int GetRandomWidth()
        {
            return rand.Next(0, WIDTH / 10);
        }

        private int GetRandomHeight()
        {
            return rand.Next(0, HEIGHT / 10);
        }
    }
}

File: Java/Flyweight/src/DrawingForm.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class DrawingForm extends JFrame {
    private final int WIDTH = 818;
    private final int HEIGHT = 395;
    private static final ShapeType[] shapes = {ShapeType.LINE, ShapeType.OVAL_FILL, ShapeType.OVAL_NOFILL};
    private static final Color[] colors = {Color.RED, Color.GREEN, Color.ORANGE};
    private final Random rand = new Random();
    private boolean stopRequested = false;
    private boolean isDrawing = false;
    private final List<IShape> shapeList = new ArrayList<>();
    public static DrawingForm Instance;

    private final JPanel pictrBox;
    private final JTextField textBoxOutput;
    private final Timer drawTimer;

    public DrawingForm() {
        setTitle("Drawing Form");
        setLayout(new BorderLayout());
        setSize(WIDTH, HEIGHT);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        pictrBox = new JPanel() {
            @Override
            protected void paintComponent(Graphics g) {
                super.paintComponent(g);
                if (stopRequested || shapeList.size() >= 20) {
                    clearMessage();
                    isDrawing = false;
                    return;
                }

                for (IShape shape : shapeList) {
                    shape.draw(g);
                }
            }
        };

        JButton btnStart = new JButton("Start");
        JButton btnStop = new JButton("Stop");
        btnStart.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                if (!isDrawing) {
                    stopRequested = false;
                    shapeList.clear();
                    isDrawing = true;
                    drawTimer.start();
                }
            }
        });
        btnStop.addActionListener(e -> stopRequested = true);

        textBoxOutput = new JTextField();
        drawTimer = new Timer(2000, e -> {
            if (!stopRequested && shapeList.size() < 20) {
                IShape shape = ShapeFactory.getShape(getRandomShape(), getRandomColor(), getRandomX(), getRandomY(), getRandomWidth(), getRandomHeight());
                shapeList.add(shape);
                pictrBox.repaint();
            }
        });

        JPanel controlsPanel = new JPanel();
        controlsPanel.add(btnStart);
        controlsPanel.add(btnStop);

        add(pictrBox, BorderLayout.CENTER);
        add(textBoxOutput, BorderLayout.SOUTH);
        add(controlsPanel, BorderLayout.NORTH);

        Instance = this;
    }

    public void clearMessage() {
        textBoxOutput.setText("");
    }

    public void showMessage(String message) {
        textBoxOutput.setText(textBoxOutput.getText() + message);
    }

    private ShapeType getRandomShape() {
        return shapes[rand.nextInt(shapes.length)];
    }

    private Color getRandomColor() {
        return colors[rand.nextInt(colors.length)];
    }

    private int getRandomX() {
        return rand.nextInt(WIDTH);
    }

    private int getRandomY() {
        return rand.nextInt(HEIGHT);
    }

    private int getRandomWidth() {
        return rand.nextInt(WIDTH / 10);
    }

    private int getRandomHeight() {
        return rand.nextInt(HEIGHT / 10);
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new DrawingForm().setVisible(true));
    }
}